### PR TITLE
feat(support): fetch agent network info over tailnet

### PR DESCRIPTION
**Pull request by johnstcn**
Part of https://github.com/coder/coder/issues/12161

Adds following information to support bundle:
- Agent
- Listening Ports
- Manifest from agent (in-memory)
- Peer Diagnostics from agent
- Magicsock HTML from agent
- Agent ping result
- `/tmp/coder-agent.log` from agent

### CODE REVIEW

File: cli/support_test.go [hunk 1]
```diff
@@ -141,22 +155,44 @@ func assertBundleContents(t *testing.T, path string) {
...
+			bs := readBytesFromZip(t, f)
+			require.NotEmpty(t, bs, "agent magicsock should not be empty")
+		case "agent/manifest.json":
+			var v agentsdk.Manifest
+			decodeJSONFromZip(t, f, &v)
+			require.NotEmpty(t, v, "agent manifest should not be empty")
+		case "agent/peer_diagnostics.json":
+			var v *tailnet.PeerDiagnostics
+			decodeJSONFromZip(t, f, &v)
+			require.NotEmpty(t, v, "peer diagnostics should not be empty")
```
Comment (mafredri) [inline cli/support_test.go:190]: How well does NotEmpty work with pointers? Should it be dereferenced perhaps?

File: support/support.go [hunk 2]
```diff
@@ -296,6 +279,121 @@ func WorkspaceInfo(ctx context.Context, client *codersdk.Client, log slog.Logger
...
+				a.MagicsockHTML = msBytes
+				return nil
+			})
+
+			eg.Go(func() error {
+				manifestRes, err := conn.DebugManifest(ctx)
+				if err != nil {
+					return xerrors.Errorf("fetch manifest: %w", err)
+				}
+				if err := json.NewDecoder(bytes.NewReader(manifestRes)).Decode(&a.Manifest); err != nil {
```
Comment (mafredri) [inline support/support.go:363]: Wondering, is there a need to decode this, actually? Or should we just use the bytes? Do we actually access any of the data here, or do we reformat it?

File: support/support.go [hunk 3]
```diff
@@ -49,8 +54,17 @@ type Workspace struct {
 	TemplateVersion    codersdk.TemplateVersion           `json:"template_version"`
 	TemplateFileBase64 string                             `json:"template_file_base64"`
 	BuildLogs          []codersdk.ProvisionerJobLog       `json:"build_logs"`
-	Agent              codersdk.WorkspaceAgent            `json:"agent"`
-	AgentStartupLogs   []codersdk.WorkspaceAgentLog       `json:"startup_logs"`
+}
+
+type Agent struct {
+	Agent           codersdk.WorkspaceAgent                        `json:"agent"`
+	ListeningPorts  *codersdk.WorkspaceAgentListeningPortsResponse `json:"listening_ports"`
```
Comment (mafredri) [inline support/support.go:61]: Why do we use a ptr here? (Same could perhaps be asked for e.g. manifest.)

File: support/support.go [hunk 4]
```diff
@@ -296,6 +279,121 @@ func WorkspaceInfo(ctx context.Context, client *codersdk.Client, log slog.Logger
...
+		defer closer.Close()
+		var logs []codersdk.WorkspaceAgentLog
+		for logChunk := range agentLogCh {
+			logs = append(logs, logChunk...)
+		}
+		a.StartupLogs = logs
+		return nil
+	})
+
+	dialCtx, dialCancel := context.WithCancel(ctx)
```
Comment (mafredri) [inline support/support.go:316]: I don't think this needs a separate context.

File: support/support.go [hunk 5]
```diff
@@ -346,6 +444,11 @@ func Run(ctx context.Context, d *Deps) (*Bundle, error) {
 		b.Network = ni
 		return nil
 	})
+	eg.Go(func() error {
+		ai := AgentInfo(ctx, d.Client, d.Log, d.AgentID)
```
Comment (mafredri) [inline support/support.go:448]: Such ai, much smart!

File: support/support_test.go [hunk 6]
```diff
@@ -24,14 +29,18 @@ import (
 	"github.com/coder/coder/v2/testutil"
 )
 
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(m)
```
Comment (mafredri) [inline support/support_test.go:33]: ❤️

**Review by mafredri — APPROVED** [review]
Forgot the approve button.

File: support/support.go (hunk 2 above)
Comment (johnstcn) [reply support/support.go:363]: We could probably just write the bytes. We can still validate in the tests that it is, in fact, a valid manifest.

**Review by dannykopping — APPROVED** [review]
LGTM

File: support/support.go (hunk 2 above)
Comment (johnstcn) [reply support/support.go:363]: I forgot -- we do actually reformat for readability 😅 So keeping as-is.

File: support/support.go (hunk 5 above)
Comment (johnstcn) [reply support/support.go:446]: wow

File: cli/support_test.go (hunk 1 above)
Comment (johnstcn) [reply cli/support_test.go:190]: Added a test helper to assert not nil and not empty.

**Pull request merged**